dir.c

```diff
@@ -1387,25 +1387,37 @@ static enum path_treatment treat_directory(struct dir_struct *dir,
 		return path_none;
 
 	case index_nonexistent:
-		if (dir->flags & DIR_SHOW_OTHER_DIRECTORIES)
-			break;
+		if (dir->flags & DIR_SHOW_OTHER_DIRECTORIES) {
+			if (!(dir->flags & DIR_HIDE_EMPTY_DIRECTORIES))
+				return exclude ? path_excluded : path_untracked;
+
+			untracked = lookup_untracked(dir->untracked, untracked,
+				     dirname + baselen, len - baselen);
+			return read_directory_recursive(dir, istate, dirname, len,
+					untracked, 1, 0, pathspec);
+		}
```

Comment:
I think it is overall easier to reason about if all the logic follows a consistent pattern (which I think my change accomplishes). If we want to minimize the diff, I think suggestion would accomplish that... I don't feel strongly - I think either way works.